### PR TITLE
[WK2] Use Spans when encoding, decoding ArrayReferenceTuple objects

### PR DESCRIPTION
**Pull request by zdobersek**
#### b67edb25e4326f0b505d2aaa22062d9e13a75e4b
<pre>
[WK2] Use Spans when encoding, decoding ArrayReferenceTuple objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=249727">https://bugs.webkit.org/show_bug.cgi?id=249727</a>

Reviewed by NOBODY (OOPS!).

This adjusts the ArgumentCoder specialization for ArrayReferenceTuple
to use Span objects for encoding and decoding, along with some other
tweaks.

In encoding, Spans are adopted and used in a fold-expression over the
encoder&apos;s stream-insertion operator.

In decoding, the decoded size is retrieved and fitted into size_t. Then,
another fold-expression is used to detect any overflows when the size
value is multiplied by the byte-size for each type. Before, this was
done for each decoded tuple element separately. Data for each tuple
element is now decoded as a Span, corresponding to the new way of
encoding.

* Source/WebKit/Platform/IPC/ArgumentCoders.h:
</pre>
https://github.com/WebKit/WebKit/commit/b67edb25e4326f0b505d2aaa22062d9e13a75e4b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101220 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10373 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34273 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110505 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11310 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1240 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93624 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108339 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107003 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8606 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91848 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/35152 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90508 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23254 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78149 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4014 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24768 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4064 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1197 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10169 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44243 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5818 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebKit/Platform/IPC/ArgumentCoders.h [hunk 1]
```diff
@@ -124,39 +124,39 @@ struct ArgumentCoder<ArrayReferenceTuple<Types...>> {
...
 
-        if constexpr (Index < sizeof...(Types)) {
-            using ElementType = ArrayReferenceTupleElementType<Index>;
-            auto dataSize = CheckedSize { size } * sizeof(ElementType);
-            if (UNLIKELY(dataSize.hasOverflowed()))
-                return std::nullopt;
-            const uint8_t* data = decoder.decodeFixedLengthReference(dataSize, alignof(ElementType));
-            if (UNLIKELY(!data))
+        if constexpr (index < sizeof...(Types)) {
+            auto data = decoder.template decode<Span<const ArrayReferenceTupleElementType<index>>>();
```
Comment (kkinnunen-apple) [inline Source/WebKit/Platform/IPC/ArgumentCoders.h:154]: So if `ArrayReferenceTupleElementType<1>` is `uint32_t`, does this not invoke `decoder.template decode<Span<const uint32_t>`?

Above in `struct ArgumentCoder<Span<T, Extent>>` the decode function
-  decodes `uint64_t` for size (which has to match the size we already know)
- decodes the data

Where as compare on the old revision on the left side <--

The intention of the code before was to lay out the data like:
- size
- data of type1 (`size*sizeof(type1)` bytes)
- data of type2 (`size*sizeof(type2)` bytes)
- ...

If I understand correctly, the current proposed encoding changes this to:
- size
- size, data of type 1
- size, data of type 2
- ...

The problem with this is that it encodes redundant info and always then rises the question that "what if the individual sizes don't match? Where do we handle that? Where do we test that we handle that?".

The change here could prove that the encoding does not change by adding an unit test..

File: Source/WebKit/Platform/IPC/ArgumentCoders.h [hunk 2]
```diff
@@ -124,39 +124,39 @@ struct ArgumentCoder<ArrayReferenceTuple<Types...>> {
...
+    static std::optional<ArrayReferenceTuple<Types...>> decode(Decoder& decoder, size_t size, DataPointerTypes&&... dataPointers)
     {
-        constexpr size_t Index = sizeof...(DataPointerTypes);
-        static_assert(Index <= sizeof...(Types));
+        constexpr size_t index = sizeof...(DataPointerTypes);
+        static_assert(index <= sizeof...(Types));
 
-        if constexpr (Index < sizeof...(Types)) {
-            using ElementType = ArrayReferenceTupleElementType<Index>;
-            auto dataSize = CheckedSize { size } * sizeof(ElementType);
```
Comment (kkinnunen-apple) [inline Source/WebKit/Platform/IPC/ArgumentCoders.h:151]: In the left side, we only decode:
 - decode the data (we know the size already)

**Comment by kkinnunen-apple** [comment]
Considering definition of encoding primitive type is something that when encoded, decode, the coder takes the type as byte array and calls encode/decodeFixedLengthData.

If I understand correctly, the target of the change is to reduce the amount of encoding primitive types.
Before:
 - {u}int{8,16,32,64}_t, ...
 - ArrayReference (aka Span)
 - ArrayReferenceTuple

After:
 - {u}int{8,16,32,64}_t, ...
 - ArrayReference (aka Span)

So it's not clear from the commit message why exactly this is desirable. E.g. is it solving a problem or is it just so to make it simpler by reducing the primitives.

**Comment by zdobersek** [comment]
Yes, this isn't good, there's added overhead. I'll revisit it after #7967.

**Comment by zdobersek** [comment]
Superseded by #8672.

**Pull request closed without merging**